lib/firestore/delete.js

```diff
@@ -280,18 +289,100 @@ FirestoreDelete.progressBar = new ProgressBar("Deleted :current docs (:rate docs
  */
 FirestoreDelete.prototype._recursiveBatchDelete = function() {
   var self = this;
-  return this._getDescendantBatch(this.allDescendants, this.batchSize).then(function(docs) {
-    if (docs.length <= 0) {
-      return Promise.resolve();
+
+  // Tunable deletion parameters
+  var maxOutstanding = 150;
+  var minimumQueueSize = 150;
+
+  // All temporary variables for the deletion queue.
+  var queue = [];
+  var numOutstanding = 0;
+  var pagesRemaining = true;
+  var pageIncoming = false;
+  var lastDocName = undefined;
+
+  var failures = [];
+
+  var queueLoop = function() {
+    if (queue.length == 0 && numOutstanding == 0 && !pagesRemaining) {
+      return true;
     }
 
-    return self._deleteDocuments(docs).then(function(numDeleted) {
-      // Tick the progress bar
-      FirestoreDelete.progressBar.tick(numDeleted);
+    if (queue.length <= minimumQueueSize && pagesRemaining && !pageIncoming) {
+      pageIncoming = true;
 
-      // Recurse to delete another batch
-      return self._recursiveBatchDelete();
-    });
+      self
+        ._getDescendantBatch(self.allDescendants, self.batchSize, lastDocName)
+        .then(function(docs) {
+          pageIncoming = false;
+
+          if (docs.length == 0) {
+            pagesRemaining = false;
+            return;
+          }
+
+          docs.forEach(function(doc) {
+            queue.push(doc);
+          });
+
+          lastDocName = docs[docs.length - 1].name;
+        })
+        .catch(function(e) {
+          logger.debug("Failed to fetch page after " + lastDocName, e);
+          pageIncoming = false;
+        });
+    }
+
+    if (numOutstanding > maxOutstanding) {
+      return false;
+    }
+
+    if (queue.length == 0) {
+      return false;
+    }
+
+    const toDelete = queue.shift();
+    numOutstanding++;
+
+    self
+      ._deleteDocument(toDelete)
+      .then(function() {
+        FirestoreDelete.progressBar.tick(1);
+        numOutstanding--;
+      })
+      .catch(function(e) {
+        // For server errors, just throw the document back into the queue.
+        if (e.status >= 500 && e.status < 600) {
```

Comment:
I am concerned about this since if there was an outage, then the CLI will get into an infinite loop attempting the same deletes. Can you include a retry counter in the queue and only retry once?

Reply:
Added something to only retry each doc one time.